src/main/scala/com/fulcrumgenomics/bam/api/SamOrder.scala

```diff
@@ -0,0 +1,221 @@
+  def groupOrder: GroupOrder
+
+  /** The subsort string to be placed *after* the sort order in the SS tag. */
+  def subSort: Option[String]
+
+  /** Function to generate the sort key for sorting records into this ordering. */
+  def sortkey: (SamRecord) => A
+}
+
+object SamOrder {
+  /** The set of all possible values. */
+  val values: Seq[SamOrder] = Seq(Coordinate, Queryname, Random, RandomQuery, TemplateCoordinate, Unsorted, Unknown)
+
+  /**
+    * Returns the SamOrder for the given name. Throws an exception if the name doesn't match a known ordering.
+    * Performs case insensitive matching so that we can use capitalized class names, but have backwards
+    * compatibility to SAM spec sort orders that are all lower-case.
+    * */
+  def apply(name: String): SamOrder = values.find(_.productPrefix.equalsIgnoreCase(name)).getOrElse {
+    throw new NoSuchElementException("No SamOrder: " + name)
+  }
+
+  /** If the header represents a known sort order returns that SortOrder otherwise None.. */
+  def apply(header: SAMFileHeader): Option[SamOrder] = {
+    // TODO: Update this to allow partial matches?
+    val so = header.getSortOrder
+    val go = Option(header.getGroupOrder).getOrElse(GroupOrder.none)
+
+    // TODO: Update this so that if the SS prefix is not SO, discard SS?
+    val ss = Option(header.getAttribute("SS")).map(s => s.substring(s.indexOf(':') + 1))
+    values.find(sam => sam.sortOrder == so && sam.groupOrder == go && sam.subSort == ss)
+  }
+
+  /** Ordering object for coordinate order per the SAM spec. */
+  case object Coordinate extends SamOrder {
+    override type A = CoordinateKey
+    override val sortOrder:  SortOrder      = SortOrder.coordinate
+    override val groupOrder: GroupOrder     = GroupOrder.none
+    override val subSort:    Option[String] = None
+    override val sortkey: SamRecord => A = rec => {
+      val ref = rec.refIndex
+      val idx = if (ref < 0) Int.MaxValue else ref
+      CoordinateKey(idx, rec.start, rec.asSam.getFlags)
+    }
+  }
+
+  /** Ordered key object for Coordinate order. */
+  case class CoordinateKey(refIndex: Int, pos: Int, flags: Int) extends Ordered[CoordinateKey] {
+    override def compare(that: CoordinateKey): Int = {
+      var retval = this.refIndex - that.refIndex
+      if (retval == 0) retval = this.pos - that.pos
+      if (retval == 0) retval = this.flags - that.flags
+      retval
+    }
+  }
+
+  /** Ordering object for queryname order per the SAM spec. */
+  case object Queryname extends SamOrder {
+    override type A = QuerynameKey
+    override val sortOrder:  SortOrder      = SortOrder.queryname
+    override val groupOrder: GroupOrder     = GroupOrder.none
+    override val subSort:    Option[String] = None
+    override val sortkey: (SamRecord => A)  = r => QuerynameKey(r.name, r.asSam.getFlags)
+  }
+
+  /** Ordered key object for Queryname order. */
+  case class QuerynameKey(name: String, flags: Int) extends Ordered[QuerynameKey] {
```

Comment:
It _could_ but it gets quite messy because Java only really supports unsigned `Int` and `Long`, and not `Short`.  So I could store it as a `Short`, but then have to use methods to convert it to an unsigned int at every use.  Not clear that would be a win.